### PR TITLE
ENH - remove scaling by alpha in dual point

### LINKED ISSUE #237: ENH remove scaling by alpha in dual point
It's more canon to use `theta = R` than `theta = R / alpha` and it will make the code simpler

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
closes #237

### CODE REVIEW

**Comment by Badr-MOUFAD** [comment]
@mathurinm, can you have a look at the last commit? 
I might have forgotten something.

File: celer/PN_logreg.pyx [hunk 1]
```diff
@@ -110,7 +110,8 @@ def newton_celer(
             is_sparse, theta, X, X_data, X_indices, X_indptr,
             screened, X_mean, weights_pen, center, positive)
 
-        if norm_Xtheta > 1.:
+        # alpha instead of 1
+        if norm_Xtheta > alpha:
```
Comment (mathurinm) [inline celer/PN_logreg.pyx:114]: if the norm is greater than alpha, you should divide by norm(X.T @ theta) / alpha, not by norm(X.T @ theta)

File: celer/group_fast.pyx [hunk 2]
```diff
@@ -334,7 +336,9 @@ cpdef celer_grp(
         for epoch in range(max_epochs):
             if epoch != 0 and epoch % gap_freq == 0:
                 fcopy(&n_samples, &R[0], &inc, &theta_inner[0], &inc)
-                tmp = 1. / (alpha * n_samples)
+
```
Comment (mathurinm) [inline celer/group_fast.pyx:339]: the diff is easier to read without comments and new lines

**Comment by Badr-MOUFAD** [comment]
What should be the tha value of ``alphas`` in ``celer_path`` if it is not provided

refer to: https://github.com/mathurinm/celer/blob/83c46284b712c42d499044ab4ef29e4fe2a0df63/celer/homotopy.py#L189

**Comment by Badr-MOUFAD** [comment]
Well noted! I will fix it rather in ``WIP - add enet support #230``
Here ``l1_ratio`` won't affect the calculation as it equals 1

File: celer/cython_utils.pyx [hunk 3]
```diff
@@ -406,10 +419,11 @@ cpdef floating dnorm_l1(
 @cython.wraparound(False)
 @cython.cdivision(True)
 cdef void set_prios(
-    bint is_sparse, floating[:] theta,
+    bint is_sparse, floating[:] theta, floating alpha,
     floating[::1, :] X, floating[:] X_data, int[:] X_indices, int[:] X_indptr,
     floating[:] norms_X_col, floating[:] weights, floating[:] prios,
     int[:] screened, floating radius, int * n_screened, bint positive) nogil:
+    # TODO pass alpha to this function
```
Comment (mathurinm) [inline celer/cython_utils.pyx:426]: rm

File: celer/cython_utils.pyx [hunk 4]
```diff
@@ -187,26 +192,34 @@ cdef floating dual_logreg(int n_samples, floating alpha, floating * theta,
...
+# remove alpha
+# [before] should work with alpha = 1
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
 cdef void create_dual_pt(
         int pb, int n_samples, floating alpha, floating * out,
         floating * R, floating * y) nogil:
     """It is scaled by alpha for both Lasso and Logreg"""
+    alpha = 1.
```
Comment (mathurinm) [inline celer/cython_utils.pyx:222]: fix it properly rather than doing that

File: celer/cython_utils.pyx [hunk 5]
```diff
@@ -187,26 +192,34 @@ cdef floating dual_logreg(int n_samples, floating alpha, floating * theta,
...
 
+# remove alpha
+# [before] should work with alpha = 1
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
 cdef void create_dual_pt(
         int pb, int n_samples, floating alpha, floating * out,
         floating * R, floating * y) nogil:
     """It is scaled by alpha for both Lasso and Logreg"""
```
Comment (mathurinm) [inline celer/cython_utils.pyx:221]: rm this comment that is no longer valid

File: celer/group_fast.pyx [hunk 6]
```diff
@@ -117,11 +117,12 @@ cpdef floating dnorm_grp(
 @cython.wraparound(False)
 @cython.cdivision(True)
 cdef void set_prios_grp(
-        bint is_sparse, int pb, floating[::1] theta, floating[::1, :] X,
+        bint is_sparse, int pb, floating[::1] theta, floating alpha, floating[::1, :] X,
         floating[::1] X_data, int[::1] X_indices, int[::1] X_indptr,
         floating[:] weights, floating[::1] norms_X_grp, int[::1] grp_ptr,
         int[::1] grp_indices, floating[::1] prios, int[::1] screened,
         floating radius, int * n_screened):
+    # TODO pass alpha to this function
```
Comment (mathurinm) [inline celer/group_fast.pyx:125]: rm

File: celer/group_fast.pyx [hunk 7]
```diff
@@ -229,15 +232,18 @@ cpdef celer_grp(
     for t in range(max_iter):
         # if t != 0: TODO potential speedup at iteration 0
         fcopy(&n_samples, &R[0], &inc, &theta[0], &inc)
-        tmp = 1. / (alpha * n_samples)
+
+        # remove scale
```
Comment (mathurinm) [inline celer/group_fast.pyx:236]: rm these comments when you do the change below

File: celer/multitask_fast.pyx [hunk 8]
```diff
@@ -65,9 +65,10 @@ cdef floating dual_scaling_mtl(
 @cython.cdivision(True)
 cdef void set_prios_mtl(
         floating[:, ::1] W, int[:] screened,
-        floating[::1, :] X, floating[::1, :] theta, floating[:] norms_X_col,
+        floating[::1, :] X, floating[::1, :] theta, floating alpha, floating[:] norms_X_col,
         floating[:] Xj_theta, floating[:] prios, floating radius,
         int * n_screened) nogil:
+    # TODO pass alpha to this function
```
Comment (mathurinm) [inline celer/multitask_fast.pyx:71]: rm

File: celer/multitask_fast.pyx [hunk 9]
```diff
@@ -190,24 +191,29 @@ def celer_mtl(
         p_obj = primal_mtl(n_samples, n_features, n_tasks, W, alpha, R)
         # theta = R / alpha:
         fcopy(&n_obs, &R[0, 0], &inc, &theta[0, 0], &inc)
-        tmp = 1. / alpha
+
+        # remove scale
+        tmp = 1.  # 1. / alpha
```
Comment (mathurinm) [inline celer/multitask_fast.pyx:196]: rm the fscal instead

File: celer/multitask_fast.pyx [hunk 10]
```diff
@@ -328,17 +334,18 @@ cpdef void inner_solver(
             p_obj = primal_mtl(n_samples, n_features, n_tasks, W, alpha, R)
             fcopy(&n_obs, &R[0, 0], &inc, &theta[0, 0], &inc)
 
-            tmp = 1. / (alpha * n_samples)
+            # remove scale
+            tmp = 1.  # / (alpha * n_samples)
```
Comment (mathurinm) [inline celer/multitask_fast.pyx:338]: why does n_samples disappear ?

**Comment by Badr-MOUFAD** [comment]
@mathurinm, there are still some tests that fail.
More precisely after debugging, I noticed when running ``celer`` with an array of ``alphas`` (such that ``len(alphas)>1``), the resulting ``coefs`` are the same and equal the ``coef`` of the first ``alpha``. ``tests/debug_lasso_test.py`` reproduces that.
Also, the unit testing on ``tests/test_mtl.py`` are veryyy slow.

**Comment by mathurinm** [comment]
If you take the same grid of alphas but fit alpha per alpha, with one call to celer path per alpha, do you get convergence ?
You can set verbose=2 to get more info

**Comment by Badr-MOUFAD** [comment]
> If you take the same grid of alphas but fit alpha per alpha, with one call to celer path per alpha, do you get convergence ? You can set verbose=2 to get more info

Yes, ``celer_path`` converges when I pass in the ``alphas`` one by one alpha

File: celer/cython_utils.pyx [hunk 11]
```diff
@@ -172,13 +174,16 @@ cdef floating dual_lasso(int n_samples, floating alpha, floating norm_y2,
     """Theta must be feasible"""
     cdef int i
     cdef floating d_obj = 0.
+    alpha = 1.
```
Comment (mathurinm) [inline celer/cython_utils.pyx:177]: do it cleanly: remove this line, and implement the formula without the scaling by alpha

File: celer/cython_utils.pyx [hunk 12]
```diff
@@ -187,26 +192,34 @@ cdef floating dual_logreg(int n_samples, floating alpha, floating * theta,
...
                    floating * theta, floating * y) nogil:
+
+    alpha = 1.
     if pb == LASSO:
         return dual_lasso(n_samples, alpha, norm_y2, &theta[0], &y[0])
     else:
         return dual_logreg(n_samples, alpha, &theta[0], &y[0])
 
 
+# remove alpha
```
Comment (mathurinm) [inline celer/cython_utils.pyx:213]: can you avoid comments like these which when you don't need them anymore, which make the diff longer to review ?

File: celer/PN_logreg.pyx [hunk 13]
```diff
@@ -377,9 +378,9 @@ cpdef int PN_logreg(
                 notin_WS, X_mean, weights_pen, center, 0)
 
         for i in range(n_samples):
-            aux[i] /= max(1, norm_Xaux)
+            aux[i] /= max(1, norm_Xaux / alpha)
```
Comment (Badr-MOUFAD) [inline celer/PN_logreg.pyx:381]: The rescaling here is not implemented like in the other files.
Yet, I assume that we should divide by ``alpha``.

Comment (mathurinm) [reply celer/PN_logreg.pyx:381]: yes, makes sense
do you identify why the build fails?

File: celer/PN_logreg.pyx [hunk 14]
```diff
@@ -105,16 +105,17 @@ def newton_celer(
         p_obj = primal(LOGREG, alpha, Xw, y, w, weights_pen)
 
         # theta = y * sigmoid(-y * Xw) / alpha
```
Comment (mathurinm) [inline celer/PN_logreg.pyx:107]:
```suggestion
        # theta = y * sigmoid(-y * Xw)
```

File: celer/PN_logreg.pyx [hunk 15]
```diff
@@ -105,16 +105,17 @@ def newton_celer(
...
-        create_dual_pt(LOGREG, n_samples, alpha, &theta[0], &Xw[0], &y[0])
+        create_dual_pt(LOGREG, n_samples, &theta[0], &Xw[0], &y[0])
         norm_Xtheta = dnorm_l1(
             is_sparse, theta, X, X_data, X_indices, X_indptr,
             screened, X_mean, weights_pen, center, positive)
 
-        if norm_Xtheta > 1.:
-            tmp = 1. / norm_Xtheta
+        # alpha instead of 1
```
Comment (mathurinm) [inline celer/PN_logreg.pyx:113]: rm

File: celer/PN_logreg.pyx [hunk 16]
```diff
@@ -105,16 +105,17 @@ def newton_celer(
...
+        if norm_Xtheta > alpha:
+            tmp = 1. / (norm_Xtheta / alpha)
```
Comment (mathurinm) [inline celer/PN_logreg.pyx:115]: probably cleaner: tmp = alpha / norm_Xtheta  (multiple occurrences)

File: celer/PN_logreg.pyx [hunk 17]
```diff
@@ -161,8 +162,8 @@ def newton_celer(
             for i in range(n_samples):
                 theta_acc[i] = y[i] * sigmoid(- y[i] * theta_acc[i])
 
-            tmp = 1. / alpha
-            fscal(&n_samples, &tmp, &theta_acc[0], &inc)
+            tmp = 1.  #  1. / alpha
```
Comment (mathurinm) [inline celer/PN_logreg.pyx:165]: rm

File: celer/cython_utils.pyx (hunk 5 above)
Comment (mathurinm) [reply celer/cython_utils.pyx:213]: Remove the docstring comment L209, because it no longer valid

File: celer/PN_logreg.pyx (hunk 16 above)
Comment (Badr-MOUFAD) [reply celer/PN_logreg.pyx:114]: Shall we adopt this in all occurrences of ``tmp``?

Comment (Badr-MOUFAD) [reply celer/PN_logreg.pyx:114]: It's true that ``alpha / norm_Xtheta`` is clean. Yet, I think ``1. / (norm_Xtheta / alpha)`` is more readable as it related to the dual equations.

Do we scarify readability for cleanability?

Comment (mathurinm) [reply celer/PN_logreg.pyx:114]: Go for alpha/norm everywhere. It is quite readable too.

**Comment by Badr-MOUFAD** [comment]
@mathurinm, the removal of ``alpha`` should also be considered in the acceleration, namely ``create_accel_pt``

https://github.com/mathurinm/celer/blob/83c46284b712c42d499044ab4ef29e4fe2a0df63/celer/cython_utils.pyx#L243-L244

**Comment by Badr-MOUFAD** [comment]
By making a comparison between the old version (main branch) and the updated version (current branch)

updated-version

old-version
)

It seems that both versions give similar results.
(refer to script ``tests/check_speed.py`` to reproduce)

**Comment by mathurinm** [comment]
can you compare the timings on not too dummy data with %timeit, on main and on this branch ?

In your Lasso fig the times are very low and this may not reflect the full picture

File: celer/multitask_fast.pyx [hunk 18]
```diff
@@ -105,6 +105,7 @@ cdef floating dual_mtl(
     cdef int i, k
     cdef floating d_obj = 0.
 
+    alpha = 1
```
Comment (mathurinm) [inline celer/multitask_fast.pyx:108]: can you remove this ?

File: celer/multitask_fast.pyx [hunk 19]
```diff
@@ -328,17 +330,16 @@ cpdef void inner_solver(
             p_obj = primal_mtl(n_samples, n_features, n_tasks, W, alpha, R)
             fcopy(&n_obs, &R[0, 0], &inc, &theta[0, 0], &inc)
 
-            tmp = 1. / (alpha * n_samples)
+            tmp = 1. / n_samples
             # tmp = 1. / alpha
```
Comment (mathurinm) [inline celer/multitask_fast.pyx:334]: can you remove this old line?

File: celer/PN_logreg.pyx [hunk 20]
```diff
@@ -161,8 +161,7 @@ def newton_celer(
             for i in range(n_samples):
                 theta_acc[i] = y[i] * sigmoid(- y[i] * theta_acc[i])
 
-            tmp = 1. / alpha
-            fscal(&n_samples, &tmp, &theta_acc[0], &inc)
+            tmp = 1.
```
Comment (mathurinm) [inline celer/PN_logreg.pyx:164]: what is the use for this ?

**Comment by mathurinm** [comment]
Timings look ok on my side for the Lasso. 1 last comment to address, you can also remove the check_speed file and LGTM

File: celer/PN_logreg.pyx (hunk 20 above)
Comment (Badr-MOUFAD) [reply celer/PN_logreg.pyx:164]: I forgot to remove it

Comment (mathurinm) [reply celer/PN_logreg.pyx:164]: ok perfect, remove it and the bench file and we merge !

**Comment by mathurinm** [comment]
Thanks @Badr-MOUFAD!

**Pull request merged**